### PR TITLE
Autopilot manual negotiation integration

### PR DESCRIPTION
**Pull request by patapufate**
# Description
This PR:
- Removes control flags from Jaus Ros Bridge and Autopilot
- Integrates the [multiplexer](https://github.com/ChrisScianna/ROS-Underwater-RnD/pull/151) for actuator (thruster velocity and fin angles)
- Integrates the change made in [the fin angle msgs](https://github.com/ChrisScianna/ROS-Underwater-RnD/pull/156).
- Because of the multiplexer, Jaus Ros Bridge publishes the thruster velocity and the fin angles continuosly in manual mode.

Discussion of this PR can be found in [Manual vs Autopilot Control Negotiation During Missions](https://github.com/ChrisScianna/ROS-Underwater-RnD/issues/135)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] The code builds clean without any errors or warnings
- [x] Unit tests are passing
- [ ] Linter tests are passing

# Test
The PR was tested on the vehicle:
- Modifying fin_angles and thruster velocity
- Changing to mission control
- Launch mission
- Changing from mission to manual mode
- Aborting Mission

### CODE REVIEW

File: cmd_actuators_mux/include/cmd_actuators_mux/cmd_actuator_mux_nodelet.h [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+  ros::NodeHandle nh;
+  ros::NodeHandle pnh;
+
+  /* Pool of topics subscribers */
+  CmdActuatorSubscribers cmd_fin_angles_subs;
+  CmdActuatorSubscribers cmd_set_rpm_subs;
+
+  /* Multiplexed command topics */
+    ros::Publisher fin_angles_output_topic_pub;
```
Comment (hidmic) [inline cmd_actuators_mux/include/cmd_actuators_mux/cmd_actuator_mux_nodelet.h:85]: @patapufate nit: fix indentation

File: autopilot/src/autopilot.cpp [hunk 2]
```diff
@@ -519,12 +496,7 @@ void AutoPilotNode::spin()
   ros::Rate r(control_loop_rate_);
   while (ros::ok())
   {
-    autopilot::AutoPilotInControl message;
-    message.auto_pilot_in_control = auto_pilot_in_control_;
-    auto_pilot_in_control_pub_.publish(message);
 
-    if (auto_pilot_in_control_)
-    {
       double roll_command = 0.;  // straight
```
Comment (hidmic) [inline autopilot/src/autopilot.cpp:500]: @patapufate nit: fix indentation

File: cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+      {
+        active_subscriber_pub.publish(fin_angles_acv_msg);
+        publish_new_subscriber = false;
+      }
+      fin_angles_output_topic_pub.publish(msg);
+    }
+    else
+      ROS_DEBUG_STREAM("RPM and fin angles values come from different source"
+                       << " - set RPM source: " << cmd_set_rpm_subs[idx]->name
+                       << " - set Fin Angles source: " << cmd_fin_angles_subs[idx]->name);
```
Comment (hidmic) [inline cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp:77]: @patapufate consider
```suggestion
                       << " - set RPM source: " << set_rpm_acv_msg.data
                       << " - set Fin Angles source: " << fin_angles_acv_msg.data);
```
which is probably more accurate.

File: cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+    // The fin angle input has changed from an input to another
+    fin_angles_acv_msg.data = cmd_fin_angles_subs[idx]->name;
+
+    // Wait until both RPM and Fin commands come from the same source
+    if (set_rpm_acv_msg.data == fin_angles_acv_msg.data)
+    {
+      if (publish_new_subscriber)
+      {
+        active_subscriber_pub.publish(fin_angles_acv_msg);
+        publish_new_subscriber = false;
```
Comment (hidmic) [inline cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp:70]: @patapufate meta: a single message on change may be problematic for late joiners.

File: fin_control/src/fin_control.cpp [hunk 5]
```diff
@@ -244,13 +242,14 @@ void FinControl::handleSetAngles(const fin_control::SetAngles::ConstPtr &msg)
   float angle_plus_offset;
 
   // check for max angle the fins can mechanically handle
-  if ((fabs(msg->f1_angle_in_radians) > maxCtrlFinAngle) ||
-      (fabs(msg->f2_angle_in_radians) > maxCtrlFinAngle) ||
-      (fabs(msg->f3_angle_in_radians) > maxCtrlFinAngle) ||
-      (fabs(msg->f4_angle_in_radians) > maxCtrlFinAngle))
+  for (uint8_t i = 0; i < 4; i++)
```
Comment (hidmic) [inline fin_control/src/fin_control.cpp:245]: @patapufate nit: 
```suggestion
  for (int i = 0; i < 4; ++i)
```

File: jaus_ros_bridge/src/SetFinDeflection.cpp [hunk 6]
```diff
@@ -60,38 +67,46 @@ void SetFinDeflection::ProcessData(char* message) {
...
-  //    count++;
-
-  //    sleep(0.05);
-  //    printf("sleeping for 50ms. count= %d \n", count);
-  //}
-
-  if (_PresenceVector->IsBitSet((int)0)) {
+
+  //  The OCU sends _finID  from 1 to 4
+  if(_finId < 5)
```
Comment (hidmic) [inline jaus_ros_bridge/src/SetFinDeflection.cpp:75]: @patapufate nit: 
```suggestion
  if (_finId <= 4)
```
is less surprising.

File: jaus_ros_bridge/src/SetFinDeflection.cpp [hunk 7]
```diff
@@ -60,38 +67,46 @@ void SetFinDeflection::ProcessData(char* message) {
...
 }
+
+void SetFinDeflection::commandFinAngles(const ros::TimerEvent& ev)
+{
+  fin_control::SetAngles msg;
+    for (int i = 0; i < 4; i++)
+    {
+      msg.fin_angle_in_radians[i] = _finAngles[i];
+    }
+    _publisher_setAngles.publish(msg);
```
Comment (hidmic) [inline jaus_ros_bridge/src/SetFinDeflection.cpp:103]: @patapufate nit: fix indentation

File: mission_control/src/mission_control.cpp [hunk 8]
```diff
@@ -278,7 +280,14 @@ void MissionControlNode::queryMissionsCallback(const mission_control::QueryMissi
...
 }
-
+void MissionControlNode::stopMissionCallback(const std_msgs::Bool& msg)
+{
+  if (current_mission_ && msg.data)
+  {
+    reportOn(current_mission_->preempt());
+    ROS_DEBUG_STREAM("stopMissionsCallback - Stopping mission");
+  }
+}
```
Comment (hidmic) [inline mission_control/src/mission_control.cpp:290]: @patapufate meta: I'm not thrilled with having some interfaces failing and some interfaces being a no-op if no mission is active. We should homogenize this later on.

**Review by hidmic — COMMENTED** [review]
@patapufate consider rebasing to remove https://github.com/ChrisScianna/ROS-Underwater-RnD/pull/160/commits/34ac7ba4919c39349f08c2245d58169325d24095, https://github.com/ChrisScianna/ROS-Underwater-RnD/pull/160/commits/6604e06c305b6a0588187a4dffc0f02aa9c21176, and https://github.com/ChrisScianna/ROS-Underwater-RnD/pull/160/commits/5c149e8b893048ca46bdacc97cc1643ece158bb0. They are already present in `master`

File: cmd_actuators_mux/include/cmd_actuators_mux/cmd_actuator_mux_nodelet.h (hunk 1 above)
Comment (patapufate) [reply cmd_actuators_mux/include/cmd_actuators_mux/cmd_actuator_mux_nodelet.h]: Fixed in 5cde148

File: autopilot/src/autopilot.cpp (hunk 2 above)
Comment (patapufate) [reply autopilot/src/autopilot.cpp:500]: Fixed in 5cde148

File: cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp (hunk 3 above)
Comment (patapufate) [reply cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp]: Fixed in 5cde148

File: fin_control/src/fin_control.cpp (hunk 5 above)
Comment (patapufate) [reply fin_control/src/fin_control.cpp]: Fixed in 5cde148

File: jaus_ros_bridge/src/SetFinDeflection.cpp (hunk 6 above)
Comment (patapufate) [reply jaus_ros_bridge/src/SetFinDeflection.cpp]: Fixed in 5cde148

File: jaus_ros_bridge/src/SetFinDeflection.cpp (hunk 7 above)
Comment (patapufate) [reply jaus_ros_bridge/src/SetFinDeflection.cpp]: Fixed in 5cde148

File: cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp (hunk 4 above)
Comment (patapufate) [reply cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp]: Changed in e1fc8af

File: cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp [hunk 9]
```diff
@@ -0,0 +1,427 @@
...
+    // waint until both, RPM and Fin commands come from the same source
+    if (set_rpm_acv_msg.data == fin_angles_acv_msg.data)
+    {
+      active_subscriber_pub.publish(set_rpm_acv_msg);
+      set_rpm_output_topic_pub.publish(msg);
+    }
+    else
+      ROS_DEBUG_STREAM("RPM and fin angles values come from different source"
+                       << " - set RPM source: " << cmd_set_rpm_subs[idx]->name
+                       << " - set Fin Angles source: " << cmd_fin_angles_subs[idx]->name);
```
Comment (hidmic) [inline cmd_actuators_mux/src/cmd_actuator_mux_nodelet.cpp:96]: @patapufate same as in lines 72-73.

File: mission_control/test/test_jaus_ros_bridge_interface_mission_control.py [hunk 10]
```diff
@@ -133,6 +134,24 @@ def executing_mission_status_is_reported():
             wait_for(executing_mission_status_is_reported),
             msg='Mission control must report EXECUTING')
 
+        # Stop mission
+        self.mission.stop_mission()
+
+        def paused_mission_status_is_reported():
```
Comment (hidmic) [inline mission_control/test/test_jaus_ros_bridge_interface_mission_control.py:140]: @patapufate nit: this predicate needs another name.

**Review by hidmic — APPROVED** [review]
LGTM

I haven't tested it myself but I trust your QA @patapufate.

File: jaus_ros_bridge/src/JausDataManager.cpp [hunk 11]
```diff
@@ -152,36 +124,20 @@ void JausDataManager::ProcessReceivedData(char* buffer)
...
-    if(!_activateManualControlEnabled&&_deactivateFromOCU){
+      _missionCommands.StopMission();
       ROS_INFO(ACTIVATE_MAUNAL_CONTROL);
-      _deactivateFromOCU = false;
-    }
-    
-    _activateManualControlEnabled = true;
-    //_needTimerUpdate = true;
+      _finControl.PublishFinAngles(true);
+      _thrusterControl.PublishRPM(true);
```
Comment (hidmic) [inline jaus_ros_bridge/src/JausDataManager.cpp:133]: @patapufate we made a mistake here. Dropping manual control timeouts in favor of the mux makes sense so long as the bridge just bridges. This is holding state. If we the connection with the OCU is lost, the bridge will keep publishing the last message and the mux won't ever switch back.

**Pull request merged**